src/lib/string_pool.rs

```diff
@@ -0,0 +1,535 @@
+        self.inner().rent(|buf| *buf.borrow().0.last().expect("Called get_last_char() when string was empty"))
+    }
+
+    /// Appends an entire C String to the current BumpVec.
+    pub(crate) unsafe fn append_c_string(&self, mut s: *const XML_Char) -> bool {
+        self.inner().rent(|vec| {
+            let mut vec = vec.borrow_mut();
+
+            while *s != 0 {
+                if !vec.append_char(*s) {
+                    return false;
+                }
+                s = s.offset(1)
+            }
+            true
+        })
+    }
+
+    /// Resets the current Bump and deallocates its contents.
+    /// The `inner` method must never be called here as it assumes
+    /// self.0 is never `None`
+    pub(crate) fn clear(&mut self) {
+        let mut inner_pool = self.0.take();
+
+        let mut bump = inner_pool.unwrap().into_head();
+
+        bump.reset();
+
+        inner_pool = Some(InnerStringPool::new(
+            bump,
+            |bump| RefCell::new(RentedBumpVec(BumpVec::new_in(&bump))),
+        ));
+
+        swap(&mut self.0, &mut inner_pool);
+    }
+
+    pub(crate) fn store_c_string(
+        &self,
+        enc: &ENCODING,
+        buf: ExpatBufRef,
+    ) -> bool {
+        self.inner().rent(|vec| {
+            let mut vec = vec.borrow_mut();
+
+            if !vec.append(enc, buf) {
+                return false;
+            }
+            if !vec.append_char('\0' as XML_Char) {
+                return false;
+            }
+
+            true
+        })
+    }
+
+    pub(crate) fn append(
+        &self,
+        enc: &ENCODING,
+        read_buf: ExpatBufRef,
+    ) -> bool {
+        self.inner().rent(|vec| vec.borrow_mut().append(enc, read_buf))
+    }
+
+    pub(crate) unsafe fn copy_c_string(
+        &self,
+        mut s: *const XML_Char,
+    ) -> Option<&mut [XML_Char]> {
+        // self.append_c_string(s);?
+        let successful = self.inner().rent(|vec| {
+            let mut vec = vec.borrow_mut();
+
+            loop {
+                if !vec.append_char(*s) {
+                    return false;
+                }
+                if *s == 0 {
+                    break;
+                }
+                s = s.offset(1);
+            }
+
+            true
+        });
+
+        if !successful {
+            return None;
+        }
+
+        Some(self.finish_string())
+    }
+
+    pub(crate) unsafe fn copy_c_string_n(
+        &self,
+        s: *const XML_Char,
+        n: c_int,
+    ) -> Option<&[XML_Char]> {
+        let successful = self.inner().rent(|vec| {
+            let mut vec = vec.borrow_mut();
+            let iter = CStrIterator(s);
+            let n = n.try_into().unwrap();
+            let cap = vec.0.capacity();
+            let new_size = vec.0.len() + n;
+
+            if new_size > cap && vec.0.try_reserve_exact(new_size).is_err() {
+                return false;
+            };
+
+            vec.0.extend(iter.take(n));
```

Comment:
I now think you don't even what `CStrIterator` even here, since the original code always copies exactly `n` characters, while this one will stop at the null terminator. You might have to revert the `CStrIterator` commit.